### ISSUE TITLE
Nanostores in SSR

### ISSUE THREAD

**Issue by AlexandrHoroshih** [ai/nanostores#80]
Hi there!

I saw news on new release, saw `allTasks` API and decided to research, how `nanostores` is compatible with SSR

Loading of Node.js template in codesandbox takes forever for some reason, so i decided to create similiar reproduce with basic one:
https://codesandbox.io/s/loving-wilson-4vp4r?file=/src/index.js
(see `index.js` and console)

Idea is the same:
1. Request happens
2. Request-handler task is added to event-loop and starts some logic and web or db requests
3. Once those web or db requests are over, state is recalculated with new data
4. State injected in the app, which then rendered to string
5. Resulting string sent as response

During all of these operations is important to isolate state that "belongs" to different requests from each other - otherwise data from one request will leak to the other

**Nanostores have two issues there:**

1. States of nanostores are shared between requests, which then leads to data prepared for one request to show in another - in result rendered html-string is not correct.

2. State of internal `allTasks` counter is also shared between all requests, which leads to weird bug: responses are not sent until **all** requests are handled, even if their data is ready.

**Possible solutions:**

1. Many state-managers require creation of new instance of store/stores for every request, like this
```ts
const ssr = async (request) => {
 const stores = initStores()
 
 await stores.startLogic(request)
 
 return renderAppToString(stores)
}
```
This will effectively isolate requests from each other, because all of them will have their own instance of stores and all pending operations also will be bounded to this instance
[Mobx](https://github.com/mobxjs/mobx) example: https://github.com/vercel/next.js/tree/master/examples/with-mobx-state-tree

2. Some other state-managers, like [`effector`](https://github.com/effector/effector), allow to create separate instance of **app state** in special way, allowing to reuse already initialized stores and connections, like this:
```ts
const ssr = async (request) => {
 // all stores and connections already defined somewhere
 const scope = fork()
 
 await allSettled(startLogic, { scope, params: request })
 
 return renderAppToString(scope)
}
```
Example: https://github.com/GTOsss/ssr-effector-next-example/blob/effector-react-form-ssr/src/pages/ssr.tsx

**Comment by AlexandrHoroshih** [comment]
I think, nanostores could utilize similiar stack-based approach like the one in `effector` 🤔
(+ React.Context and Hooks are also working like this, but in a bit diffirent situation though)

General idea is something like this
https://share.effector.dev/sr3BtOLj
```ts
// internal field
let scope = null

const makeStore = (name, initState) => ({
 // all stores will need an unique name or identifier, which should be the stable between server and client
 name,
 state: initState,
 getState() {
  // if scope is set during operation, use state from scope
  if (scope) {
   return scope[name] ?? initState
  }
  return this.state;
 },
 set(v) {
  if (scope) {
   scope[name] = v;
   return;
  }
  this.state = v;
 }
})

// scope fabric
const getScope = () => ({});

// general pattern for working with scope
// pass target scope explicitly
const runInScope = (s, cb) => {
 // set it to internal field
 scope = s;
 // run logic
 cb()
 // thanks to callstack all store api calls will see
 // which scope is set now
 // but requires additional work to handle async stuff
 
 // clear internal field
 scope = null;
}
```

This would allow to automatically collect all the states during SSR and then transfer it to the client:
```ts
// ssr
const myServerScope = getScope()
// ... run some logic on scope
const stringifiedStateOfTheApp = JSON.stringify(myServerScope)

// client
const myClientScope = JSON.parse(stringifiedStateOfTheApp)
```

And also separate state of different requests/tests:
```ts
const myTestOrSsrRequestHandler = () => {
  // each test/request gets its own isolated box for app state
 const myScope = getScope()
 ...
}
```

This is roughly (very roughly) how effector scopes work, but advantages are the same: effector is able to both isolate and automatically collect all atomic stores states without any manual boilerplate

**Comment by ai** [comment]
Yeap, Effector’s `scope` isolation is very good for SSR, when you render HTML for multiple users in the same context.

Maybe we can create some sort of the wrap around current minimalistic stores to support scope?

**Comment by AlexandrHoroshih** [comment]
I don't think that wrapper would work - if user missed some store, then it will be missed from serialization and further hydration at the client, which is no good
Probably it should be a core feature, so all APIs of the library are aware of the scopes and can properly handle it 🤔 

I don't think that it is actually matter, if server context is isolated or not - the problem of collecting all parts of decentralized state of the app is still persists, since it is needed to hydrate app state at the client somehow anyway 🤔

**Comment by droganov** [comment]
Hm, could we pass `getServerSnapshot` to the React hook to prevent render time side effects on Server?

**Comment by bernatfortet** [comment]
Any updates on using SSR? The documentation is very short and I don't get enough clarity.

My use case is to be able to hydrate the store from a Server Component in React 18 / Next.js. The underlying reason is that by hydrating I hope to avoid an extra re-render.

Thanks!

**Comment by ai** [comment]
I don’t use Next.js and we need some advanced user to do the research and optimization

**Comment by lucastnr** [comment]
I've used the following with Next.JS and it seems to work properly.
```tsx
const $atom = atom<string>('initial');

const hydratedSet = new WeakSet();
function useHydrateAtom<T>(atom: WritableAtom<T>, value: T) {
  if (!hydratedSet.has(atom)) {
    hydratedSet.add(atom);
    atom.set(value);
  }
}

export default function Page({ value }: Props) {
  const atomValue = useStore($atom);
  useHydrateAtom($atom, value);

  return <div>Atom Value: {atomValue}</div>;
}

export const getServerSideProps: GetServerSideProps<Props> = async () => {
  return {
    props: {
      value: 'value',
    },
  };
};

```

**Comment by lucastnr** [comment]
I've been using the following to hydrate atoms on Next.js with Pages Router, but it should work fine on App Router too:
```tsx
const hydratedSet = new WeakSet();

function useHydrateAtom<T>(atom: WritableAtom<T>, value: T) {
  if (!hydratedSet.has(atom)) {
    hydratedSet.add(atom);
    atom.set(value);
  }
}

const $atom = atom<string>('');

// value is coming from the server-side
export default function Page({ value }) {
  useHydateAtom($atom, value);

  return (
    // ...
  )
}
```

**Comment by lucastnr** [comment]
@bernatfortet I think the above may help you
